apollo-federation/src/sources/connect/json_selection/parser.rs

```diff
@@ -58,9 +134,43 @@ impl JSONSelection {
         }
     }
 
-    pub fn parse(input: &str) -> IResult<&str, Self> {
-        let input_span = Span::new(input);
+    // JSONSelection::parse is possibly the "most public" method in the entire
+    // file, so it's important that the method signature can remain stable even
+    // if we drastically change implementation details. That's why we use &str
+    // as the input type and a custom JSONSelectionParseError type as the error
+    // type, rather than using Span or nom::error::Error directly.
+    pub fn parse(input: &str) -> IResult<&str, Self, JSONSelectionParseError> {
+        match JSONSelection::parse_span(new_span(input)) {
+            Ok((remainder, selection)) => {
+                // To avoid exposing the implementation details of nom_locate's
+                // LocatedSpan, we report the remainder as a &str instead.
+                Ok((remainder.fragment(), selection))
+            }
```

Comment:
Can we make this `Result<Self, JSONSelection>`, using std result, and make remainder an error? Right now all the consumers have some half-baked version of returning an error anyway when there's a remainder, and if we use std Result, we can do `impl From <JSONSelectionParseError> for Message` (and every other error type) and use `?`. So something like this:

```suggestion
    pub fn parse(input: &str) -> Result<Self, JSONSelectionParseError> {
        match JSONSelection::parse_span(new_span(input)) {
            Ok((remainder, selection)) => {
                let fragment = remainder.fragment();
                if fragment.is_empty() {
                    Ok(selection)
                } else {
                    Err(JSONSelectionParseError {
                        message: "Unexpected trailing characters".to_string(),
                        fragment: fragment.to_string(),
                        offset: remainder.location_offset(),
                    })
                }
            }
```

This also keeps noms `Err` out of our public API, so we can more easily directly use `JSONSelectionParserError`.

Reply:
Good idea! https://github.com/apollographql/router/pull/6089/commits/76949bd53a152f620bebfcef13c76f9e52e44dfe

Reply:
I think I updated all the consumers (including both `apollo-federation` and `apollo-router`), unless there are any outside of this repository?